### ISSUE TITLE
JavahExecutable fails on Java10

### ISSUE THREAD

**Issue by hendriks73** [mojohaus/maven-native#17]
Java 10 got rid of the `javah` executable and replaced it with `javac -h`.

**Comment by hendriks73** [comment]
Looks like the regular maven compiler plugin (v3.7.0) can be configured to spit out the headers like so:

            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <configuration>
                    <compilerArgs>
                        <arg>-h</arg>
                        <arg>${project.build.directory}/native/include</arg>
                    </compilerArgs>
                </configuration>
            </plugin>

**Comment by ctubbsii** [comment]
The 1.0-alpha-9 version was released. That milestone should be closed and this issue bumped to the next milestone (unless it was already fixed).

**Comment by dantran** [comment]
@hendriks73 is this fixed in alpha-9?

**Comment by hendriks73** [comment]
No, it's not.

Java9 and 10 do not have a `javah` executable anymore. So the native plugin should either

- somehow execute `javac` with the appropriate flags (not sure, if that's possible)
- show a warning that `javah` is not supported anymore
- find some other programmatic way to generate headers

That aside, users can use the regular compiler plugin to produce header files, as shown above.

**Comment by crashtua** [comment]
There is one project - https://github.com/Glavo/gjavah it looks like an implementation of javah in pure java. It is mit licensed. Maybe it can be used instead of javah?

**Comment by ctubbsii** [comment]
Given that `javac -h` works even better than `javah`, I don't really see why this even needs to be maintained anymore. It's easy enough to update the `maven-compiler-plugin` [to use `-h`][1] and, if necessary, it's not too difficult to use profiles to support different Java versions to support pre- and post-Java8, which added the `-h` flag to `javac`.

[1]: https://github.com/apache/accumulo/commit/4bd3afa2c997fdcb35d54b8be32b703385f3b5c1#diff-27da03a9af4d8a1ab61b98bce67375fc

**Comment by notiocide** [comment]
This is true but `javac -h` becomes useless when working with other JVM languages.

**Comment by ryantheseer** [comment]
> Given that `javac -h` works even better than `javah`, I don't really see why this even needs to be maintained anymore. It's easy enough to update the `maven-compiler-plugin` [to use `-h`](https://github.com/apache/accumulo/commit/4bd3afa2c997fdcb35d54b8be32b703385f3b5c1#diff-27da03a9af4d8a1ab61b98bce67375fc) and, if necessary, it's not too difficult to use profiles to support different Java versions to support pre- and post-Java8, which added the `-h` flag to `javac`.

I see that we can change org.codehaus.mojo.natives.javah.JavahExecutable to use `javac -h` instead of `javah`, but I believe the command-line changed from referring to the class file with its package name to using the .java file.  Let's say I successfully branch this project, make that small change, build and install the plugin for maven, how would I call the plugin differently?

**Comment by ryantheseer** [comment]
Also, if I understand correctly, this plugin allows you to define a build artifact that builds the JNI header and the native library (.so, .dll, .dylib) in the same build call. Do I have to separate those into two steps if I use the `maven-compiler-plugin` as noted above?

**Comment by ryantheseer** [comment]
What I've done for now is to hack the `org.codehaus.mojo.natives.javah.JavahExecutable` to use `javac -h` instead of `javah`, and modified my pom.xml to use the `.java` source path for the `<javahClassName>` property instead of using the dot-separated package and class name. I can build it with JDK8 that way. Next step is to try with a newer JDK.

**Comment by mizapf** [comment]
Is this fix for JavahExecutable already available? I'm having an issue with another project that relies on the maven-native plugin.

**Comment by jbescos** [comment]
I am also interested in this fix.
